### PR TITLE
Fix sidecar proxy deadlock during BPF generation

### PR DESCRIPTION
**Pull request by rlenglet**
In some cases, the HTTP proxy in sidecar mode can't connect to Istio Pilot to be configured during regeneration, even if the policy whitelists that traffic.

This can happen especially when the first policy being applied on an endpoint contains HTTP rules. In this case, BPF generation waits for the sidecar proxy to configure the Cilium filters and to ACK the policy configuration, before the endpoint's BPF program has been installed for this endpoint. Because no BPF program is installed, all traffic keeps getting dropped, which results in the sidecar proxy not being able to connect to Istio Pilot and to get configured. The endpoint remains in this state, with all its traffic being dropped, until it the endpoint deleted or the HTTP rules are deleted from the policy.

Specifying an init policy that whitelists traffic to Istio services at L3/L4 normally solves that problem. Such an init policy allows the sidecar proxy to get configured before the endpoint's actual policy is generated. However, in some cases, the identity of an endpoint is resolved before its init policy is generated, so the first policy to be generated is the endpoint's actual policy and the init policy is ignored.

To solve this problem, refactor the BPF regeneration to configure L3+L4 connectivity and generate the BPF program first, before configuring the L7 proxies and configuring L7 policies. This way, if the policy to be enforced allows connectivity at L3/L4 to Istio services, the sidecar proxy can always get configured by Istio Pilot before L7 policies are configured.

The new BPF regeneration process is now:

1. generate the policy as done currently
2. do the complete generation, incl.:
  - modify the BPF policy map, **but do not configure any new proxy redirect (i.e. any new L4 port that is to be redirected to an L7 proxy)**; this includes removing any obsolete redirects from the BPF policy map, in effect dropping all traffic to obsolete redirects
  - send the new network policy to Envoy
  - generate the BPF program
3. delete the obsolete redirects from the Envoy + Kafka proxies
4. create the new redirects to the Envoy + Kafka proxies
5. wait for Envoy to ACK all the changes; may require to increase the proxy wait timeout because that may require waiting for the BPF policy map to be configured in order for Envoy to get configured in case it's in a sidecar container
6. modify the BPF policy map to configure the new L4 ports with proxy redirects

This requires that the BPF datapath looks up redirect ports from the BPF policy map entries (https://github.com/cilium/cilium/pull/2918) and for the Cilium agent to populate the redirect ports in policy map entries (https://github.com/cilium/cilium/issues/4711).

Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4610)

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Hand-tested that it works as intented.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
While manual testing, I hit a bug in the `controller` package:
```
Jun 22 01:05:38 runtime1 cilium-agent[13876]: panic: close of closed channel
Jun 22 01:05:38 runtime1 cilium-agent[13876]: goroutine 1502 [running]:
Jun 22 01:05:38 runtime1 cilium-agent[13876]: github.com/cilium/cilium/pkg/controller.(*Controller).stopController(...)
Jun 22 01:05:38 runtime1 cilium-agent[13876]:         /home/vagrant/go/src/github.com/cilium/cilium/pkg/controller/controller.go:251
Jun 22 01:05:38 runtime1 cilium-agent[13876]: github.com/cilium/cilium/pkg/controller.(*Manager).removeController(0xc4201bdd90, 0xc423040f70, 0xc4231f2201)
Jun 22 01:05:38 runtime1 cilium-agent[13876]:         /home/vagrant/go/src/github.com/cilium/cilium/pkg/controller/manager.go:90 +0x43
Jun 22 01:05:38 runtime1 cilium-agent[13876]: github.com/cilium/cilium/pkg/controller.(*Manager).UpdateController(0xc4201bdd90, 0xc4231f22a0, 0x13, 0xc4231495f0, 0x0, 0x0, 0x0, 0x0, 0x0)
Jun 22 01:05:38 runtime1 cilium-agent[13876]:         /home/vagrant/go/src/github.com/cilium/cilium/pkg/controller/manager.go:65 +0x47f
Jun 22 01:05:38 runtime1 cilium-agent[13876]: github.com/cilium/cilium/pkg/endpoint.(*Endpoint).UpdateLabels(0xc4201bdb80, 0x3c36440, 0xc4202667e0, 0xc423149530, 0xc423149560)
Jun 22 01:05:38 runtime1 cilium-agent[13876]:         /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2214 +0x1dc
Jun 22 01:05:38 runtime1 cilium-agent[13876]: github.com/cilium/cilium/pkg/workloads.(*dockerClient).handleCreateWorkload(0xc42000e4a8, 0xc421b0f740, 0x40, 0x126b300)
Jun 22 01:05:38 runtime1 cilium-agent[13876]:         /home/vagrant/go/src/github.com/cilium/cilium/pkg/workloads/docker.go:477 +0x876
Jun 22 01:05:38 runtime1 cilium-agent[13876]: created by github.com/cilium/cilium/pkg/workloads.(*watcherState).syncWithRuntime.func1
Jun 22 01:05:38 runtime1 cilium-agent[13876]:         /home/vagrant/go/src/github.com/cilium/cilium/pkg/workloads/watcher_state.go:133 +0xb5
```

This happened when I removed the `ErrorRetryBaseDuration` field from my `ControllerParams`. I'll set it back to 1 second.

**Comment by rlenglet** [comment]
test-me-please

**Comment by aanm** [comment]
@rlenglet did you fix the controller's bug?

**Comment by rlenglet** [comment]
@aanm no, still working on it.

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -2204,21 +2204,46 @@ func (e *Endpoint) IsInit() bool {
...
-	e.Mutex.Lock()
-	e.replaceInformationLabels(infoLabels)
-	// replace identity labels and update the identity if labels have changed
-	rev := e.replaceIdentityLabels(identityLabels)
-	e.Mutex.Unlock()
-	if rev != 0 {
-		e.runLabelsResolver(owner, rev)
-	}
+				e.Mutex.Lock()
+				if e.RealizedL4Policy == nil {
```
Comment (aanm) [inline pkg/endpoint/endpoint.go]: Will this work even if there's no L4 policy selecting the endpoint?

**Comment by aanm** [comment]
>Start a controller to delay the update of the labels until we have at least generated completely one policy for the endpoint.

What's wrong on updating the labels before generated the first policy?

**Comment by rlenglet** [comment]
@aanm The problem can occur when an L7/HTTP policy is imported, then a pod with sidecar proxy is created.

If the labels are updated before the init policy is compiled, the init policy is never in effect. Here’s what happens in this case:
- No BPF program is currently installed (no policy has been generated yet), so all traffic is dropped.
- We try to generate the pod’s actual policy. Since it has L7/HTTP rules, and the pod is in sidecar proxy mode, we wait for the sidecar proxy to ACK the policy, before we compile and install the BPF program.
- Since all traffic is still dropped, the sidecar Envoy proxy can’t connect to Istio Pilot. Therefore, it can’t get configured, so it can’t install the Cilium filters, so it can’t query and ACK the policy configuration from the Cilium agent.
- Since the agent never receives an ACK from Envoy, the policy generation fails.
- This situation can’t be fixed. We never install any BPF program. All traffic keeps getting dropped.

By waiting until we at least compile the policy for the init policy, before we update the labels, we will at least allow Envoy to connect to Istio Pilot and get configured, so subsequent policy generations can happen.

File: pkg/endpoint/endpoint.go (hunk 1 above)
Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: @aanm ok, then what is a reliable signal that we’ve had at least one successful policy generation?

Comment (aanm) [reply pkg/endpoint/endpoint.go]: `<-e.WaitForPolicyRevision(ctx, 1)`?

Comment (aanm) [reply pkg/endpoint/endpoint.go]: Wait, that will not work. This on the other hand has been waiting for you

https://github.com/cilium/cilium/blob/adf4e4f5b31f6d020969aeea1a0a45c6af270cc8/pkg/endpoint/endpoint.go#L372-L376

Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: Beautiful!

Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: Actually, I think `PolicyCalculated` won't work, because this doesn't indicate whether a BPF program has been installed. This is the signal I need.

Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: Checking that `e.policyRevision > 0` seems to be the most reliable.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
I nailed the controller bug. I significantly cleaned up the design in the process.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Successful CI build:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4405/

**Comment by tgraf** [comment]
@rlenglet How is this problem specific to the init phase? Isn't this a generic problem that we can't install an L7 problem one the policy does not allow communication with the Istio control plane? The same will apply if a user loads a regular L3/L4 only policy which does not allow communication to pilot and then loads an L7 policy.

**Comment by rlenglet** [comment]
@tgraf Yes, it's specific to the init phase. In your case, it's a user error. The problem I'm solving is quite different. Even if the policies given by users (init policy and actual policy) correctly allow traffic to Istio services (i.e. no user error in this case), we might still prevent the proxy from getting configured, because we drop all traffic until the first policy is completely generated.

File: pkg/endpoint/endpoint.go [hunk 2]
```diff
@@ -2204,21 +2204,44 @@ func (e *Endpoint) IsInit() bool {
...
-	e.Mutex.Lock()
-	e.replaceInformationLabels(infoLabels)
-	// replace identity labels and update the identity if labels have changed
-	rev := e.replaceIdentityLabels(identityLabels)
-	e.Mutex.Unlock()
-	if rev != 0 {
-		e.runLabelsResolver(owner, rev)
-	}
+				e.Mutex.Lock()
+				if e.policyRevision == 0 {
```
Comment (ianvernon) [inline pkg/endpoint/endpoint.go]: Can you document for policyRevision within the Endpoint type that the zero value represents that no policy has been generated?

File: pkg/policy/repository.go [hunk 3]
```diff
@@ -39,7 +39,9 @@ type Repository struct {
 
 // NewPolicyRepository allocates a new policy repository
 func NewPolicyRepository() *Repository {
-	return &Repository{}
+	return &Repository{
+		revision: 1,
```
Comment (ianvernon) [inline pkg/policy/repository.go]: Add a comment indicating why we are starting at revision one.

**Review by ianvernon — COMMENTED** [review]
Nothing blocking, mainly documentation. Controller code looks much simpler now !

File: pkg/endpoint/endpoint.go (hunk 2 above)
Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: Done.

File: pkg/policy/repository.go (hunk 3 above)
Comment (rlenglet) [reply pkg/policy/repository.go]: Done.

**Comment by rlenglet** [comment]
@ianvernon I addressed your comments.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
CI build failed because of infra:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4428/execution/node/69/log/
```
05:18:53 Timed out while waiting for the machine to boot. This means that
05:18:53 Vagrant was unable to communicate with the guest machine within
05:18:53 the configured ("config.vm.boot_timeout" value) time period.
05:18:53 
05:18:53 If you look above, you should be able to see the error(s) that
05:18:53 Vagrant had when attempting to connect to the machine. These errors
05:18:53 are usually good hints as to what may be wrong.
05:18:53 
05:18:53 If you're using a custom box, make sure that networking is properly
05:18:53 working and you're able to connect to the machine. It is a common
05:18:53 problem that networking isn't setup properly in these boxes.
05:18:53 Verify that authentication configurations are also setup properly,
05:18:53 as well.
05:18:53 
05:18:53 If the box appears to be booting properly, you may want to increase
05:18:53 the timeout ("config.vm.boot_timeout") value.
```

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Review by ianvernon — DISMISSED** [review]

**Comment by tgraf** [comment]
@rlenglet I still don't understand the difference between init and non-init.

1. Istio is deployed
2. Pods are started with injection enabled
3. L3 only policy is loaded
4. L3-L7 policy is loaded which also includes access granted to istio-system
5. Deadlock

**Comment by rlenglet** [comment]
@tgraf Your case can be fixed by modifying the policy. The case I'm fixing here can only be fixed by either disabling policy enforcement, or restarting the pod.

**Comment by rlenglet** [comment]
@tgraf In your case, the policy loaded in step 3. is incorrect (assuming that you meant that it doesn't allow egress to istio-system). In my case, all the policies (the init policy as well as the actual pod's policy) are correct. They all allow egress to istio-system. But we still prevent the sidecar proxy to get initialized in some cases, and we keep dropping all traffic, even though the policies are correct.

**Comment by rlenglet** [comment]
@tgraf And in your case, we would typically NOT deadlock. Assuming that egress to istio-system is allowed before step 3., i.e. we have a chance to install a policy that allows egress to istio-system, the proxy can get initialized and get at least one Cilium filter installed. As long as at least one Cilium filter has been installed, for ingress or egress traffic, Envoy will ACK policies from Cilium, and we wouldn't deadlock.

The problem I'm fixing here, is that in some cases we don't even have a chance to install even one Cilium filter, because Envoy can't get configured at all.

**Comment by rlenglet** [comment]
Reviewed 3 of 5 files at r2, 1 of 4 files at r4, 3 of 3 files at r5.
Review status: all files reviewed, 3 unresolved discussions (waiting on @aanm and @ianvernon)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4610)*

**Comment by tgraf** [comment]
How can a policy be "incorrect"? The user loads a number of policy yamls, one of them is the L3 egress to allow to istio-system, another policy might be loaded first and will be generated on the endpoint. The user might also remove and re-add the policy and we will re-generate endpoints in between. Are you assuming that the policy allowing to istio-system is always preloaded before any pod with injection and is never removed?

File: pkg/endpoint/endpoint.go [hunk 4]
```diff
@@ -2204,21 +2206,46 @@ func (e *Endpoint) IsInit() bool {
...
-	// replace identity labels and update the identity if labels have changed
-	rev := e.replaceIdentityLabels(identityLabels)
-	e.Mutex.Unlock()
-	if rev != 0 {
-		e.runLabelsResolver(owner, rev)
-	}
+				e.Mutex.Lock()
+				// If any policy has been completely realized for this
+				// endpoint, its policyRevision is always > 0.
+				if e.policyRevision == 0 {
```
Comment (tgraf) [inline pkg/endpoint/endpoint.go]: How does this behave if policy is disabled or if no policy is loaded? policyRevision will never go above > 0?

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -130,7 +130,7 @@ type Controller struct {
 	lastErrorStamp    time.Time
```
Comment (tgraf) [inline pkg/controller/controller.go]: Please submit this commit as separate PR so we can backport if something else depends on this.

**Review by tgraf — DISMISSED** [review]

**Comment by rlenglet** [comment]
If the user wants to use Cilium with Istio sidecars, the user must import policies that allow egress DNS and egress TCP to istio-system. We can't synthesize those policies for the user.

File: pkg/endpoint/endpoint.go (hunk 4 above)
Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: The initial, empty repository starts with revision 1. So if/when a policy is generated from the initial empty repository, the endpoint will have its `policyRevision` set to 1.
Every regeneration will result in setting `policyRevision` to a value >0.

File: pkg/endpoint/endpoint.go [hunk 6]
```diff
@@ -2204,21 +2206,46 @@ func (e *Endpoint) IsInit() bool {
...
+					e.Mutex.Unlock()
+					scopedLog.Debug("Deferring update of endpoint labels until a policy has been realized")
+					// If no policy has been realized yet, continue trying.
+					return errors.New("no policy has been realized yet")
```
Comment (tgraf) [inline pkg/endpoint/endpoint.go]: This is not ideal, it will lead to failing controllers in a completely normal scenario. We will count and expose these as failing controllers in the metrics.

File: pkg/endpoint/endpoint.go [hunk 7]
```diff
@@ -2204,21 +2206,46 @@ func (e *Endpoint) IsInit() bool {
...
+				// replace identity labels and update the identity if labels have changed
+				rev := e.replaceIdentityLabels(identityLabels)
+				e.Mutex.Unlock()
+				if rev != 0 {
+					e.runLabelsResolver(owner, rev)
+				}
+
+				return nil
+			},
+			ErrorRetryBaseDuration: 1 * time.Second,
```
Comment (tgraf) [inline pkg/endpoint/endpoint.go]: Note that this will perform a constant backoff so you might delay the setting of the labels for several seconds

**Comment by tgraf** [comment]
> If the user wants to use Cilium with Istio sidecars, the user must import policies that allow egress DNS and egress TCP to istio-system. We can't synthesize those policies for the user.

I understand and agree but you make further assumptions. You assume that the no L7 policy is ever loaded before access to the istio-system has been implemented. I'm not sure how we can expect this if we allow deletion and re-addition of policies.

We can safely apply and commit all L3/L4 policies which do not contain L7 rules in a initial phase before applying all other policies to solve the deadlock. It will fix the problem throughout the lifecycle of policies.

**Comment by rlenglet** [comment]
> I understand and agree but you make further assumptions. You assume that the no L7 policy is ever loaded before access to the istio-system has been implemented. I'm not sure how we can expect this if we allow deletion and re-addition of policies.

That is not true. I only assume that the init policy doesn't contain L7 HTTP rules.

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -168,90 +168,103 @@ func (c *Controller) GetLastErrorTimestamp() time.Time {
...
+		if runFunc {
+			interval = params.RunInterval
+
+			err = params.DoFunc()
+
+			c.mutex.Lock()
+
+			if err != nil {
+				switch err := err.(type) {
+				case ExitReason:
```
Comment (raybejjani) [inline pkg/controller/controller.go]: We ended up not using this anywhere. If it makes your PRs simpler, we can remove the exit behaviour outright.

**Comment by tgraf** [comment]
> That is not true. I only assume that the init policy doesn't contain L7 HTTP rules.

@rlenglet OK, then please walk me through the following two use cases. It is still unclear to me:

1) User runs Istio and Cilium in sidecar mode without any Cilium policies loaded at first and then loads L3-L7 policies which includes a policy to enable egress to DNS and istio-system as well as L7 policies.
2) User deletes and re-adds the policy allowing egress to istio-system while an L7 policy is in place.

**Comment by rlenglet** [comment]
Successful CI build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4466/

File: pkg/endpoint/endpoint.go (hunk 6 above)
Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: Yeah, I think I will need to make the controller periodic with a non-zero interval and return an ExitReason when it succeeds. That will avoid the interval increasing backoff as well.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Discussed with @tgraf. I'll go with a different design of `regenerateBPF`:

1. generate the policy as done currently
2. do the complete generation, incl.:
  - modify the BPF policy map, **but do not configure any new proxy redirect (i.e. any new L4 port that is to be redirected to an L7 proxy)**
  - send the new network policy to Envoy
  - send the new redirects to the Envoy + Kafka proxies
  - generate the BPF program
3. delete the obsolete redirects from the Envoy + Kafka proxies
4. wait for Envoy to ACK all the changes; may require to increase the proxy wait timeout because that may require waiting for the BPF policy map to be configured in order for Envoy to get configured in case it's in a sidecar container
5. modify the BPF policy map to configure the new L4 ports with proxy redirects

This requires that the BPF datapath looks up redirect ports from the BPF policy map entries (https://github.com/cilium/cilium/pull/2918) and for the Cilium agent to populate the redirect ports in policy map entries (https://github.com/cilium/cilium/issues/4711).

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Please review this new implementation!

**Comment by rlenglet** [comment]
CI build failure due to provisioning:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4757/
```
00:32:53 K8sValidatedPolicyTest [BeforeEach]
00:32:53 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
00:32:53   Basic Test
00:32:53   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
00:32:53     checks all kind of Kubernetes policies
00:32:53     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
00:32:53 
00:32:53     
00:32:53             ===================== ERROR - VM PROVISION FAILED =====================
00:32:53     
00:32:53             Unable to provision and start VM "k8s1-1.10": exit status 1", vm, err
00:32:53     
00:32:53             =======================================================================
00:32:53             
00:32:53 
00:32:53     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:460
```

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Successful CI build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4758/
Running another build.

**Comment by rlenglet** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build link: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4761/

Build failure is due to an error provisioning the VM:

```
==> k8s1-1.10: Running provisioner: ipv6-config (shell)...
    k8s1-1.10: Running: inline script
==> k8s1-1.10: Running provisioner: shell...
    k8s1-1.10: Running: inline script
==> k8s1-1.10: Running provisioner: file...
==> k8s1-1.10: Running provisioner: shell...
    k8s1-1.10: Running: /tmp/vagrant-shell20180707-2293-1zdusb.sh
    k8s1-1.10: updating /etc/resolv.conf
    k8s1-1.10: beginning trying up to 10 times function "wget https://packages.cloud.google.com/apt/doc/apt-key.gpg"
    k8s1-1.10: ***************************************
    k8s1-1.10: *
    k8s1-1.10: *
    k8s1-1.10: * on attempt 0 of function "wget https://packages.cloud.google.com/apt/doc/apt-key.gpg"
    k8s1-1.10: *
    k8s1-1.10: *
    k8s1-1.10: ***************************************
    k8s1-1.10: --2018-07-07 03:33:07--  https://packages.cloud.google.com/apt/doc/apt-key.gpg
    k8s1-1.10: Resolving packages.cloud.google.com (packages.cloud.google.com)... 
    k8s1-1.10: 172.217.10.110, 2607:f8b0:4006:811::200e
    k8s1-1.10: Connecting to packages.cloud.google.com (packages.cloud.google.com)|172.217.10.110|:443... 
    k8s1-1.10: connected.
    k8s1-1.10: HTTP request sent, awaiting response... 
    k8s1-1.10: 200 OK
    k8s1-1.10: Length: 1326 (1.3K) [application/octet-stream]
    k8s1-1.10: Saving to: 'apt-key.gpg'
    k8s1-1.10: 
    k8s1-1.10:      0K .
    k8s1-1.10:                                                      
    k8s1-1.10: 100%  617K=0.002s
    k8s1-1.10: 2018-07-07 03:33:07 (617 KB/s) - 'apt-key.gpg' saved [1326/1326]
    k8s1-1.10: running of "wget https://packages.cloud.google.com/apt/doc/apt-key.gpg" successful
    k8s1-1.10: Warning: apt-key output should not be parsed (stdout is not a terminal)
    k8s1-1.10: OK
    k8s1-1.10: Disabling crio CNI bridge
    k8s1-1.10: removed '/etc/cni/net.d/100-crio-bridge.conf'
    k8s1-1.10: removed '/etc/cni/net.d/200-loopback.conf'
    k8s1-1.10: Reading package lists...
    k8s1-1.10: E: Could not get lock /var/lib/apt/lists/lock - open (11: Resource temporarily unavailable)
    k8s1-1.10: E: Unable to lock directory /var/lib/apt/lists/
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

Trying again.

**Comment by ianvernon** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Comment by rlenglet** [comment]
Successful CI build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4763/

File: pkg/endpoint/bpf.go [hunk 9]
```diff
@@ -331,26 +331,33 @@ func (e *Endpoint) addNewRedirectsFromMap(owner Owner, m policy.L4PolicyMap, des
...
-
-			redirectPort, err := owner.UpdateProxyRedirect(e, &l4, proxyWaitGroup)
-			if err != nil {
-				return err
-			}
+			var redirectPort uint16
+			var err error
+			// Only create a redirect if the proxy is NOT running in a sidecar
+			// container. If running in a sidecar container, just allow traffic
+			// to the port at L4 by setting the proxy port to 0.
```
Comment (aanm) [inline pkg/endpoint/bpf.go:338]: This question might be out of scope, but how does the traffic gets redirected into the sidecar?

File: pkg/endpoint/bpf.go [hunk 10]
```diff
@@ -331,26 +331,33 @@ func (e *Endpoint) addNewRedirectsFromMap(owner Owner, m policy.L4PolicyMap, des
...
+			if !e.hasSidecarProxy || l4.L7Parser != policy.ParserTypeHTTP {
```
Comment (aanm) [inline pkg/endpoint/bpf.go:339]: Different question. What will be the behavior for ongoing connections?

**Review by aanm — DISMISSED** [review]

File: pkg/endpoint/bpf.go (hunk 9 above)
Comment (rlenglet) [reply pkg/endpoint/bpf.go:338]: Istio injects an init container into the pod spec. The init container inserts an iptables TPROXY rule that redirects all TCP connections to the sidecar proxy. That's done once for the whole pod and doesn't change at runtime.

File: pkg/endpoint/bpf.go (hunk 10 above)
Comment (rlenglet) [reply pkg/endpoint/bpf.go:339]: The sidecar proxy mode is determined once when the pod is created, and doesn't change at runtime. So we don't switch between sidecar mode and non-sidecar mode.

Comment (rlenglet) [reply pkg/endpoint/bpf.go:339]: So we don't switch between redirect and non-direct mode for a given policy.
Actually, in sidecar proxy mode, we never have any issues with existing connections, between we never redirect in Cilium, and Istio configures iptables in the pod to always redirect to the proxy unconditionally.

Comment (aanm) [reply pkg/endpoint/bpf.go:339]: So, istio does the policy enforcement from L3 up to L7?

File: pkg/endpoint/bpf.go (hunk 9 above)
Comment (aanm) [reply pkg/endpoint/bpf.go:338]: Who is in charge of doing policy enforcement for UDP?

File: pkg/endpoint/bpf.go (hunk 10 above)
Comment (rlenglet) [reply pkg/endpoint/bpf.go:339]: In our specific integration with Cilium, we enforce L3 and L4 in BPF, and enforce L7 using our Envoy filter in Istio’s sidecar Envoy proxy.

File: pkg/endpoint/bpf.go (hunk 9 above)
Comment (rlenglet) [reply pkg/endpoint/bpf.go:338]: Cilium. Istio does nothing about traffic that is not TCPv4.
